### ISSUE TITLE
Teams video recording audio out of sync

### ISSUE THREAD

**Issue by doucheface1** [microsoftgraph/microsoft-graph-comms-samples#339]
we have the policy recording bot up and running and we have audio and desktop sharing recoding perfectly

the problem is video

when we record video the audio and video are out of sync we have the audio and video raw files but what ever we use we cannot get them to consistently stay in sync 

they play in sync if we use the audiovideoframeplayer but this tool does not allow use to save the output to a file.

any ideas 

thanks

**Comment by 1fabi0** [comment]
Is the not sync distance always the same like the audio is always a second to early or is it always a different distance? and if the unsync is always the same what is earlier? audio or Video?

**Comment by doucheface1** [comment]
its always different and gradually over time gets more out of sync
we think its something to do with FPS changing within video and being different for different users

**Comment by 1fabi0** [comment]
I would recommend you to save the incoming images and have a timer in the app that takes in the right time, like when you want a 30fps video output, 30 times per second the last image received and add it to the raw file,then the framer ate should be constant

**Comment by 1fabi0** [comment]
Update

So yes the Framerate of videos can adapt during streaming due to the sender Internet connection or bot's internet connection(if it's not in a data center) but the bot get notified when streaming quality changes with a special event in the VideoSocket also with the new FrameRate etc. So you may have to also save the changing Framerates and process your recordings after the Meeting, so this one is definitely not a bug so it can be closed

**Comment by stefan2410** [comment]
@1fabi0    which event?
>when streaming quality changes with a special event in the VideoSocket also with the new FrameRate

**Comment by 1fabi0** [comment]
@stefan2410 see [this](https://microsoftgraph.github.io/microsoft-graph-comms-samples/docs/bot_media/Microsoft.Skype.Bots.Media.IVideoSocket.html#Microsoft_Skype_Bots_Media_IVideoSocket_MediaStreamQualityChanged) one in the documentation

**Comment by stefan2410** [comment]
@1fabi0 
The VideoKeyFrameNeededEventArgs   does  not provide information about the  new FrameRate

**Comment by 1fabi0** [comment]
Sorry it was the wrong Link [this](https://microsoftgraph.github.io/microsoft-graph-comms-samples/docs/bot_media/Microsoft.Skype.Bots.Media.IVideoSocket.html#Microsoft_Skype_Bots_Media_IVideoSocket_MediaStreamQualityChanged) one is the correct

**Comment by 1fabi0** [comment]
Also every received video frame should provide information about the frame rate

**Comment by stefan2410** [comment]
@1fabi0   and the  MediaQualityState or the VideoQualityType  of the VideoMediaStreamQualityChangedEventArgs   don't provide information  for the frameRate
>Also every received video frame should provide information about the frame rate
We know that  and it is valid only for  the  keyframe 
But we would like to avoid  parsing each buffer

**Comment by 1fabi0** [comment]
OK but if quality changes you need a new key frame in h264 so you don't need to know whether it's a key frame or not because when it changes to a valid number you get a new key frame

**Comment by 1fabi0** [comment]
Another solution would be receiving in nv12 or rgb24 and reencode the video to h264 or even h265 then you can ensure having a constant frame rate

**Comment by stefan2410** [comment]
@1fabi0   nv12 or rgb24 are huge. not applicable for us.
btw, do you know when we subscribe for a lower resolution the the Teams sent,  the Media platform re-encode the h.264 stream?
by default the Teams server estimate the user's capabilities and sends lower if need be. IF we ask even  lower, should this re-encoidng happens in the Media platform (we don't want it)

**Comment by mosoftwareenterprises** [comment]
@doucheface1 did you every resolve this problem? 
I am experiencing the same thing, as I cannot sync video and audio.

I do not see that  FrameRate is anything other than 0 in the videreceived event.
Also, I never get a VideoMediaQuality event fired, so cannot act on that.

I am interrogating the Quality values at the end of the call to get the average frames per second, but again, that is not always available.

Any thoughts?

**Comment by 1fabi0** [comment]
@mosoftwareenterprises you can get a valid average framerate, always, if you use the timestamps of the frames you receive and calculate from duration and amount of frames the average framerate. Also with timestamps in theory you could reassembel the Video with the correct timings. Timing the raw h264 video frames correctly in a video is not an easy task easy task, as the whole system is optimized for real time media where you decode frames as soon as you receive them, you have to watch out that the order and timings between to frames are correct and you can ignore any framerate. To replay such videos, the teams/skype media sdk already takes care that the order of frames is correct at receiving but you would also need to do the replay based on the timestamps and get correct timings between frames. There are Video formats that support such things like WebM/Matroska and i think mp4 also supports timestamp values for frames, but if your video player or video format does not support this you again have a problem. Then you would need to convert from a timing based video to a framerate based video, this can be done by repeating the picture of some frames or/and deleting some pictures from frames(With pictures I mean decoded Frames), to create the framerate based video you could also take a average framerate over the whole time of the video, which is probably the easiest, but then you might have audio and video out of sync issues, because there might be a very high framerate at start and a very low framerate at the end of the video.

I hope you understand the complexity of this problem and that it might require quite a lot of work to implement a solution that solves this.

Best Regards 
Fabian

**Comment by mosoftwareenterprises** [comment]
H Fabian (@1fabi0 ) Thanks for the in-depth explanation.
I understand what you are saying.
So to reconstitute a video stream that is being recorded with the correct playback rate would require using the frames supplied at the specific timestamp they are sent. We can not assume that the next frame received should be immediately played after the previous one. This would only be true if the timestamp was correct.

So we cannot simply store all the video frames in order, and then play them back in the same order, assuming an average framerate calculated from the timestamps, this will always give incorrect results, especially when the framerate changes during the call.

Thanks, I understand the problem now, as you say, this is a lot more effort.

Thanks

Mark

**Comment by 1fabi0** [comment]
@mosoftwareenterprises yes for me you're summary sounds like it could be a bit confusing, the order of the frames is correct, that's something the media sdk is taking care of, also the gap between two frames is correct, so the time gap between two calls of the OnFrameReceived Event is also correct from the sdk. But to replay this gaps it gets complicated as I said. 

So another solution you could do is to write the video directly to a video file with ffmpeg and use clock timing for the frames, that could already be quite close to the original played video. I know that this is possible with pipes and ffmpeg but I can't give documentation for this as I think you would have to search all required documents together.

**Comment by Oreneo** [comment]
Has anything changed with this? I am experiencing the same issue

Is there a chance someone from MSFT will take a look at this?
